service/token.go

```diff
@@ -123,3 +133,80 @@ func (m *Token_Info) SetSessionKey(key []byte) {
 func (m *Token) SetSignature(sig []byte) {
 	m.Signature = sig
 }
+
+// Returns byte slice that is used for creation/verification of the token signature.
+func verificationTokenData(token SessionToken) []byte {
+	var sz int
+
+	id := token.GetID()
+	sz += id.Size()
+
+	ownerID := token.GetOwnerID()
+	sz += ownerID.Size()
+
+	verb := uint32(token.GetVerb())
+	sz += 4
+
+	addr := token.GetAddress()
+	sz += addr.CID.Size() + addr.ObjectID.Size()
+
+	cEpoch := token.CreationEpoch()
+	sz += 8
+
+	fEpoch := token.ExpirationEpoch()
+	sz += 8
+
+	key := token.GetSessionKey()
+	sz += len(key)
+
+	data := make([]byte, sz)
+
+	var off int
+
+	tokenEndianness.PutUint32(data, verb)
+	off += 4
+
+	tokenEndianness.PutUint64(data[off:], cEpoch)
+	off += 8
+
+	tokenEndianness.PutUint64(data[off:], fEpoch)
+	off += 8
+
+	off += copy(data[off:], id.Bytes())
+	off += copy(data[off:], ownerID.Bytes())
+	off += copy(data[off:], addr.CID.Bytes())
+	off += copy(data[off:], addr.ObjectID.Bytes())
+	off += copy(data[off:], key)
+
+	return data
+}
+
+// SignToken calculates and stores the signature of token information.
+//
+// If passed token is nil, ErrEmptyToken returns.
+// If passed private key is nil, crypto.ErrEmptyPrivateKey returns.
+func SignToken(token SessionToken, key *ecdsa.PrivateKey) error {
+	if token == nil {
+		return ErrEmptyToken
+	} else if key == nil {
+		return crypto.ErrEmptyPrivateKey
+	}
+
+	sig, err := crypto.Sign(key, verificationTokenData(token))
+	if err != nil {
+		return err
+	}
+
+	token.SetSignature(sig)
+
+	return nil
+}
+
+// VerifyTokenSignature checks if token was signed correctly.
+func VerifyTokenSignature(token SessionToken, key *ecdsa.PublicKey) error {
```

Comment:
`SignToken` function has nil check of `SessionToken`. I think this check will be appropriate there too.

Reply:
Undoubtedly, thx. I'll add the nil-check and test assert.